File: src/main/java/com/jackmeng/stl/stl_UnitTC.java
package com.jackmeng.stl;

public class stl_UnitTC
        implements Runnable
{
    private final String name;
    private final stl_Callback< ?, ? > action;
    private Integer expects;

    public stl_UnitTC(String name, stl_Callback< ?, ? > action)
    {
        this.name = name;
        this.action = action;
    }

    public stl_UnitTC(String name, stl_Callback< ?, ? > action, Integer expect)
    {
        this(name, action);
        expects = expect;
    }

    @Override
    public void run()
    {
        System.out.println("Testcase: " + name + "\n" + action.hashCode());
        long s = Runtime.getRuntime().maxMemory()
                - (Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()) / 1024 / 1024,
                l = System.currentTimeMillis();
        Object returns;
        try
        {
            returns = action.call((null));
        } catch (Exception e)
        {
            System.out.println(new stl_AnsiMake(new stl_AnsiColors[] { stl_AnsiColors.GREEN_BG, stl_AnsiColors.BLACK_TXT },
                    new Object[] { "FAILED ON EXCEPTION" }));
            e.printStackTrace();
            return;
        }
        long t = System.currentTimeMillis() - l;
        System.out.println("Finished in: " + t + "ms\nMemory used: "
                + (Math.abs((Runtime.getRuntime().maxMemory()
                - ((double) Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()) / 1024
                / 1024)
                - s))
                + " Mb"
                + (expects == null ? "\nCase Status: " + new stl_AnsiMake(
                new stl_AnsiColors[] { stl_AnsiColors.BLUE_BG, stl_AnsiColors.BLACK_TXT },
                new Object[] { "NONE" })
                : "\nCase status: " + ((returns.equals(expects)
                ? new stl_AnsiMake(new stl_AnsiColors[] { stl_AnsiColors.GREEN_BG, stl_AnsiColors.BLACK_TXT },
                new Object[] { "PASSED" })
                : new stl_AnsiMake(new stl_AnsiColors[] { stl_AnsiColors.RED_BG, stl_AnsiColors.BLACK_TXT },
                new Object[] { "FAILED" }))
                + "\nExpected: " + expects + "\nReturned: " + returns)));
    }
}

File: src/main/java/com/jackmeng/stl/stl_UnitTC.kt
// Copyright 2023 Jack Meng. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
package com.jackmeng.stl

import kotlin.math.abs
import java.lang.Runnable
import java.lang.Exception

class stl_UnitTC(private val name:String , private val action:stl_Callback<* , *>):Runnable
{
	private var expects:Any?=null
	
	constructor(name:String , action:stl_Callback<Any? , *> , expect:Any?):this(name , action)
	{
		expects=expect
	}
	
	override fun run()
	{
		println(
			"""
	Testcase: $name
	${action.hashCode()}
	""".trimIndent()
		)
		val s=(Runtime.getRuntime().maxMemory()-((Runtime.getRuntime().totalMemory()-Runtime.getRuntime()
			.freeMemory())/1024/1024))
		val l=System.currentTimeMillis()
		val returns:Any=try
		{
			action.call(null)
		} catch (e:Exception)
		{
			println(
				stl_AnsiMake(
					arrayOf(
						stl_AnsiColors.GREEN_BG , stl_AnsiColors.BLACK_TXT
					) , arrayOf("FAILED ON EXCEPTION")
				)
			)
			e.printStackTrace()
			return
		}
		val t=System.currentTimeMillis()-l
		println(
			"""
	Finished in: ${t}ms
	Memory used: 
	""".trimIndent()+abs(
				(Runtime.getRuntime().maxMemory()-((Runtime.getRuntime().totalMemory().toDouble()-Runtime.getRuntime()
					.freeMemory())/1024/1024))-s
			)+" Mb"+if (expects==null) "\nCase Status: "+stl_AnsiMake(
				arrayOf(
					stl_AnsiColors.BLUE_BG , stl_AnsiColors.BLACK_TXT
				) , arrayOf("NONE")
			)
			else "\nCase status: "+((if (returns==expects) stl_AnsiMake(
				arrayOf(
					stl_AnsiColors.GREEN_BG , stl_AnsiColors.BLACK_TXT
				) , arrayOf("PASSED")
			)
			else stl_AnsiMake(
				arrayOf(
					stl_AnsiColors.RED_BG , stl_AnsiColors.BLACK_TXT
				) , arrayOf("FAILED")
			)).toString()+"\nExpected: "+expects+"\nReturned: "+returns)
		)
	}
}